File: mobil/lib/core/event_timing.dart
import '../models/event.dart';

bool isPastEvent({DateTime? startAt, DateTime? endAt}) {
  final now = DateTime.now();
  if (endAt != null) {
    return endAt.isBefore(now);
  }
  if (startAt != null) {
    final dayEnd = DateTime(
      startAt.year,
      startAt.month,
      startAt.day,
      23,
      59,
      59,
      999,
    );
    return dayEnd.isBefore(now);
  }
  return false;
}

bool isPastEventModel(Event event) =>
    isPastEvent(startAt: event.startAt, endAt: event.endAt);

/// Yaklaşanlar önce, günü geçmişler sonda.
List<Event> sortEventsForDisplay(List<Event> events) {
  final copy = [...events];
  copy.sort((a, b) {
    final aPast = isPastEventModel(a);
    final bPast = isPastEventModel(b);
    if (aPast != bPast) return aPast ? 1 : -1;

    final aTs = a.startAt?.millisecondsSinceEpoch ?? 1 << 62;
    final bTs = b.startAt?.millisecondsSinceEpoch ?? 1 << 62;
    return aPast ? bTs.compareTo(aTs) : aTs.compareTo(bTs);
  });
  return copy;
}

File: frontend/src/lib/event-timing.ts
export function isPastEvent(event: {
  startAt: string | null;
  endAt?: string | null;
}): boolean {
  const now = new Date();
  if (event.endAt) {
    return new Date(event.endAt) < now;
  }
  if (event.startAt) {
    const dayEnd = new Date(event.startAt);
    dayEnd.setHours(23, 59, 59, 999);
    return dayEnd < now;
  }
  return false;
}

/** Yaklaşanlar önce (tarihe göre), günü geçmişler sonda. */
export function sortExploreEvents<
  T extends { startAt: string | null; endAt?: string | null },
>(events: T[]): T[] {
  return [...events].sort((a, b) => {
    const aPast = isPastEvent(a);
    const bPast = isPastEvent(b);
    if (aPast !== bPast) return aPast ? 1 : -1;

    const aTs = a.startAt
      ? new Date(a.startAt).getTime()
      : Number.MAX_SAFE_INTEGER;
    const bTs = b.startAt
      ? new Date(b.startAt).getTime()
      : Number.MAX_SAFE_INTEGER;

    return aPast ? bTs - aTs : aTs - bTs;
  });
}
